maven-caching/src/main/java/org/apache/maven/caching/ProjectUtils.java

```diff
@@ -0,0 +1,200 @@
+
+        return PHASES.indexOf( phase ) > PHASES.indexOf( other );
+    }
+
+    public static boolean isBuilding( Dependency dependency, ProjectIndex projectIndex )
+    {
+        final MavenProject key = new MavenProject();
+        key.setGroupId( dependency.getGroupId() );
+        key.setArtifactId( dependency.getArtifactId() );
+        key.setVersion( dependency.getVersion() );
+        return projectIndex.getProjects().containsKey( BuilderCommon.getKey( key ) );
+    }
+
+    public static boolean isPomPackaging( MavenProject project )
+    {
+        return project.getPackaging().equals( "pom" ) && !new File( getSrcDir( project ) ).exists();
+    }
+
+    public static boolean isPom( Artifact artifact )
+    {
+        return artifact.getType().equals( "pom" );
+    }
+
+    public static boolean isPom( Dependency dependency )
+    {
+        return dependency.getType().equals( "pom" );
+    }
+
+    public static boolean isSnapshot( String version )
+    {
+        return version.endsWith( SNAPSHOT_VERSION ) || version.endsWith( LATEST_VERSION );
+    }
+
+    public static String getTargetDir( MavenProject project )
+    {
+        return FilenameUtils.concat( project.getBasedir().getAbsolutePath(), "target" );
+    }
+
+    public static String getSrcDir( MavenProject project )
+    {
+        return FilenameUtils.concat( project.getBasedir().getAbsolutePath(), "src" );
+    }
+
+    public static String normalizedName( Artifact artifact )
+    {
+
+        if ( artifact.getFile() == null )
+        {
+            return null;
+        }
+
+        StringBuilder filename = new StringBuilder( artifact.getArtifactId() );
+
+        if ( artifact.hasClassifier() )
+        {
+            filename.append( "-" ).append( artifact.getClassifier() );
+        }
+
+        final ArtifactHandler artifactHandler = artifact.getArtifactHandler();
+        if ( artifactHandler != null && StringUtils.isNotBlank( artifactHandler.getExtension() ) )
+        {
+            filename.append( "." ).append( artifactHandler.getExtension() );
+        }
+        return filename.toString();
+    }
+
+    public static String mojoExecutionKey( MojoExecution mojo )
+    {
+        return StringUtils.join( Arrays.asList( StringUtils.defaultIfEmpty( mojo.getExecutionId(), "emptyExecId" ),
+                StringUtils.defaultIfEmpty( mojo.getGoal(), "emptyGoal" ),
+                StringUtils.defaultIfEmpty( mojo.getLifecyclePhase(), "emptyLifecyclePhase" ),
+                StringUtils.defaultIfEmpty( mojo.getArtifactId(), "emptyArtifactId" ),
+                StringUtils.defaultIfEmpty( mojo.getGroupId(), "emptyGroupId" ),
+                StringUtils.defaultIfEmpty( mojo.getVersion(), "emptyVersion" ) ), ":" );
+    }
+
+    public static Path getMultimoduleRoot( MavenSession session )
+    {
+        return session.getRequest().getMultiModuleProjectDirectory().toPath();
+    }
+
+    public static Scm readGitInfo( MavenSession session ) throws IOException
```

Comment:
guess we need to wire it to our scm abstraction somehow to handle svn etc...?

Reply:
I can't find any way to obtain the current branch / revision using `maven-scm`...

Reply:
Doesnt info works from ScmProvider?

Reply:
Yes, it contains the revision, not the branch.

Reply:
I guess I can reuse something like https://github.com/mojohaus/buildnumber-maven-plugin/blob/master/src/main/java/org/codehaus/mojo/build/CreateMojo.java#L626-L763

Reply:
Yep, can be worth pushing up in the API maybe?